### PR TITLE
.github: Add workflow to build libtorch

### PR DESCRIPTION
**Pull request by samestep**
Based on #50633 and #51243.

Things left to do:

- [x] modify `.github/scripts/generate_binary_build_matrix.py` further
  - [x] add option for not iterating over Python version
  - [x] add `LIBTORCH_VARIANT`
  - [x] add option for cxx11
  - [x] fix the artifact uploading
  - [x] remove `pull_request` hook before merging

**Test plan:**

[CI](https://github.com/pytorch/pytorch/actions/runs/665781075).

### CODE REVIEW

File: .github/scripts/generate_binary_build_matrix.py [hunk 1]
```diff
@@ -57,58 +77,82 @@
...
     return False
     # return os.environ.get("GITHUB_HEAD_REF")
 
-def generate_matrix(include_cuda=True, include_rocm=False):
-    python_versions = FULL_PYTHON_VERSIONS
-    cuda_versions = CUDA_ARCHES
-    rocm_versions = ROCM_ARCHES
+
+def snip_if(is_pr: bool, l: List[str]) -> List[str]:
+    return [l[-1]] if is_pr else l
```
Comment (seemethere) [inline .github/scripts/generate_binary_build_matrix.py:86]: This looks a bit confusing, I think you should change the `l` parameter to be something a bit more descriptive.

Perhaps a docstring would work as well?

**Review by seemethere — COMMENTED** [review]
Looks mostly good, just one minor nit about a particular variable

File: .github/scripts/generate_binary_build_matrix.py [hunk 2]
```diff
@@ -41,14 +51,27 @@
 
 CONDA_CONTAINER_IMAGES = {
     **{
-        # TODO: Re-do manylinux CUDA image tagging scheme to be similar to
-        #       ROCM so we don't have to do this replacement
```
Comment (walterddr) [inline .github/scripts/generate_binary_build_matrix.py:45]: did we fix this todo?

File: .github/scripts/generate_binary_build_matrix.py [hunk 3]
```diff
@@ -57,58 +80,89 @@
...
-            )
-        })
-    return json.dumps({"include": matrix})
-
-def main():
-    include_cuda = True
-    include_rocm = True
-    if os.environ.get("NO_ROCM", False):
-        include_rocm = False
-    if os.environ.get("NO_CUDA", False):
```
Comment (walterddr) [inline .github/scripts/generate_binary_build_matrix.py:104]: are we still going to support these?

**Review by walterddr — COMMENTED** [review]
got 2 qq. please kindly take a look

File: .github/scripts/generate_binary_build_matrix.py (hunk 2 above)
Comment (samestep) [reply .github/scripts/generate_binary_build_matrix.py:45]: this is to be addressed later, not in this PR

File: .github/scripts/generate_binary_build_matrix.py (hunk 3 above)
Comment (samestep) [reply .github/scripts/generate_binary_build_matrix.py:104]: these env vars? no; they're encoded in the various `generate_*_matrix` functions

File: .github/scripts/generate_binary_build_matrix.py (hunk 2 above)
Comment (walterddr) [reply .github/scripts/generate_binary_build_matrix.py:45]: hmm maybe we shouldn't remove this TODO comment then?

Comment (samestep) [reply .github/scripts/generate_binary_build_matrix.py:45]: oh nvm sorry, I didn't realize you were commenting on a TODO that was removed; yes we did resolve this

**Review by seemethere — APPROVED** [review]

**Pull request closed without merging**